drivers/kscan/kscan_matrix_poll.c

```diff
@@ -0,0 +1,307 @@
+		for (int r = 0; r < CONFIG_KSCAN_MATRIX_MAX_ROWS; r++) {
+			uint8_t mask = BIT(r);
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask)) {
+				continue;
+			}
+
+			/* Convert the clock cycle differences to usec */
+			uint32_t debt = k_cyc_to_us_floor32(
+				cycles_now - kb_poll.scan_clk_cycle[kb_poll.scan_cycle_idx[c][r]]);
+
+			/* Does the key requires more time to be debounced? */
+			if (debt < (row_bit ? kb_poll.deb_time_press : kb_poll.deb_time_rel)) {
+				/* Need more time to debounce */
+				continue;
+			}
+
+			kb_poll.matrix_unstable_state[c] &= ~row_bit;
+
+			/* Check if there was a change in the stable state */
+			if ((kb_poll.matrix_stable_state[c] & mask) == row_bit) {
+				/* Key state did not change */
+				continue;
+			}
+
+			/* The current row has been debounced, therefore update
+			 * the stable state. Then, proceed to notify the
+			 * application about the keys pressed.
+			 */
+			kb_poll.matrix_stable_state[c] ^= mask;
+			kb_poll.config.callback(dev, r, c, row_bit ? true : false);
+		}
+	}
+
+	return key_pressed;
+}
+
+void polling_task(void *dummy1, void *dummy2, void *dummy3)
+{
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (true) {
+		kscan_resume_matrix_detection(kscan_dev, true);
+		kscan_drive_matrix_column(kscan_dev, KEYBOARD_COLUMN_DRIVE_ALL);
+		k_sem_take(&poll_lock, K_FOREVER);
+		LOG_DBG("Start KB scan!!");
+
+		/* Start polling */
+		kscan_resume_matrix_detection(kscan_dev, false);
+
+		uint64_t poll_time_end = sys_clock_timeout_end_calc(K_USEC(kb_poll.poll_timeout));
+
+		while (atomic_get(kb_poll.config.enable_scan) == 1U) {
+			uint32_t start_period_cycles = k_cycle_get_32();
+
+			if (check_key_events(kscan_dev)) {
+				poll_time_end =
+					sys_clock_timeout_end_calc(K_USEC(kb_poll.poll_timeout));
+
+			} else if (k_uptime_ticks() > poll_time_end) {
+				break;
+			}
+
+			/* Subtract the time invested from the sleep period
+			 * in order to compensate for the time invested
+			 * in debouncing a key
+			 */
+			current_cycles = k_cycle_get_32();
+			cycles_diff = current_cycles - start_period_cycles;
+			wait_period = kb_poll.poll_period - k_cyc_to_us_floor32(cycles_diff);
+
+			/* Override wait_period in case it is less than 1 ms */
+			if (wait_period < USEC_PER_MSEC) {
+				wait_period = USEC_PER_MSEC;
+			}
+
+			/* wait period results in a larger number when
+			 * current cycles counter wrap. In this case, the
+			 * whole poll period is used
+			 */
+			if (wait_period > kb_poll.poll_period) {
+				LOG_DBG("wait_period : %u", wait_period);
+
+				wait_period = kb_poll.poll_period;
+			}
+
+			/* Allow other threads to run while we sleep */
+			k_usleep(wait_period);
+		}
+	}
+}
+
+static void kscan_ksi_isr_callback(const struct device *dev)
+{
+	kscan_resume_matrix_detection(kscan_dev, false);
```

Comment:
This statement is duplicated by `polling_task()` after it does `k_sem_take()`.

Reply:
>I don't see a need for the isr_callback. The only user is kscan_matrix_poll.c. You could have the drivers directly notify the matrix task.

This is intended to do it. 
I assumed that the kscan driver is unware about how the polling task react to the KSI's level change.
For example, this Zephyr's  ksan matrix polling task uses k_sem_give/k_sem_task mechanism to start/wait for the scanning.
The Cros EC's keyboard_scan_task uses the [task_wake(TASK_ID_KEYSCAN)](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3486597/6/zephyr/drivers/cros_kb_raw/cros_kb_raw.c#80) (i. e. k_polling_signal*) mechanism to do it.

>This statement is duplicated by polling_task() after it does k_sem_take().

Done by removing this statemeent.